cnxarchive/search.py

```diff
@@ -539,81 +415,132 @@ def _build_search(structured_query, weights):
     # terms do not only consist of stopwords, then use the text terms
     # without stopwords
     if arguments or text_terms_wo_stopwords:
         text_terms = text_terms_wo_stopwords
     arguments.update({'text_terms': ' '.join(text_terms)})
-    queries = '\nUNION ALL\n'.join([q for q in query_list if q is not None])
-
-    # Add the arguments for filtering.
-    filter_list = []
-    if structured_query.filters:
-        if len(filter_list) == 0:
-            filter_list.append('')  # For SQL AND joining.
-        for i, (keyword, value) in enumerate(structured_query.filters):
-            arg_name = "{}_{}".format(keyword, i)
-            # These key values are special in that they don't,
-            #   directly translate to SQL fields and values.
-            try:
-                filter_stmt, match_value = _transmute_filter(keyword, value)
-            except ValueError:
-                del structured_query.filters[i]
-                continue
-            if filter_stmt:
-                arguments[arg_name] = match_value
-                filter_stmt = filter_stmt.format(arg_name)
-                filter_list.append(filter_stmt)
-    filters = ' AND '.join(filter_list)
-
-    limits = ''
-    groupby = ''
-    having_list = []
-    having = ''
-    subject_filters = [v for k, v in structured_query.filters
-                       if k == 'subject']
-    if subject_filters:
-        limits = 'NATURAL LEFT JOIN moduletags NATURAL LEFT JOIN tags'
-        groupby = '''GROUP BY lm.name, lm.uuid, lm.portal_type, lm.authors,
-             lm.major_version, lm.minor_version, language, lm.revised,
-             ab.abstract, weight, rank, lm.module_ident, weighted.keys'''
-
-        for subj in subject_filters:
-            having_list.append("'{}' = ANY(array_agg(tag))".format(subj))
-        having = 'HAVING ' + ' AND '.join(having_list)
-    groupby = '\n'.join((groupby, having,))
-
-    if not queries:  # all filter term case
-        key_list = []
-        for key, value in structured_query.filters:
-            key_list.append(QUERY_FIELD_PAIR_SEPARATOR.join((value, key,)))
-        keys = QUERY_FIELD_ITEM_SEPARATOR.join(key_list)
-        queries = """\
-SELECT module_ident, 1 AS weight, '{}'::text AS keys
-FROM latest_modules""".format(keys)
+    # raise Exception(structured_query.filters, structured_query.terms)
+
+    if len(text_terms) > 0:
+        conditions['text_terms'] = 'AND module_idx \
+                                    @@ plainto_tsquery(%(text_terms)s)'
+
+    # build fulltext keys
+    fulltext_key = []
+    for term in text_terms_wo_stopwords:
+        fulltext_key.append(term + '-::-fulltext')
+    arguments.update({'fulltext_key':  ';--;'.join(fulltext_key)})
+
+    idx = 0
+    invalid_filters = []
+    filters = _convert(structured_query.filters, [])
+
+    while idx < len(filters):
+        keyword = filters[idx][0]
+        value = filters[idx][1]
+        if keyword == 'pubYear':
```

Comment:
this long if/elif tree isn't ideal, but is at least clear as to what's going on. Note that several branches of the tree do slightly different things …